### PR TITLE
implement `diag_indices` and `diag_indices_from`

### PR DESCRIPTION
**Pull request by rushabh-v**

### CODE REVIEW

**Comment by rushabh-v** [comment]
both the functions had their dependencies pre implemented. So there is nothing to be optimized in my view.

File: cupy/indexing/insert.py [hunk 1]
```diff
@@ -118,3 +118,72 @@ def fill_diagonal(a, val, wrap=False):
...
+    (array([0, 1]), array([0, 1]), array([0, 1]))
+    And use it to set the diagonal of an array of zeros to 1:
+    >>> a = cupy.zeros((2, 2, 2), dtype=int)
+    >>> a[d3] = 1
+    >>> a
+    array([[[1, 0],
+            [0, 0]],
+           [[0, 0],
+            [0, 1]]])
+     .. seealso:: :func:`numpy.diag_indices`
```
Comment (emcastillo) [inline cupy/indexing/insert.py]: Separate the notes with one line break

Comment (emcastillo) [reply cupy/indexing/insert.py]: Also add one line break before the closing

```
    array([[[1, 0],
            [0, 0]],
           [[0, 0],
            [0, 1]]])

     .. seealso:: :func:`numpy.diag_indices`

    """
```

File: cupy/indexing/insert.py [hunk 2]
```diff
@@ -118,3 +118,72 @@ def fill_diagonal(a, val, wrap=False):
...
+
+
+def diag_indices_from(arr):
+    """
+    Return the indices to access the main diagonal of an n-dimensional array.
+    See `diag_indices` for full details.
+
+    Args:
+        arr (cupy.ndarray): At least 2-D.
+     .. seealso:: :func:`numpy.diag_indices_from`
```
Comment (emcastillo) [inline cupy/indexing/insert.py]: ditto

File: tests/cupy_tests/indexing_tests/test_insert.py [hunk 3]
```diff
@@ -148,3 +148,37 @@ def test_1darray(self, xp, dtype):
         a = testing.shaped_arange(5, xp, dtype)
         val = self._compute_val(xp)
         xp.fill_diagonal(a, val=val, wrap=self.wrap)
+
+
+@testing.parameterize(*testing.product({
+    'n': [2, 4],
+    'ndim': [2, 3],
+}))
```
Comment (emcastillo) [inline tests/cupy_tests/indexing_tests/test_insert.py]: Test invalid values too

File: tests/cupy_tests/indexing_tests/test_insert.py [hunk 4]
```diff
@@ -148,3 +148,37 @@ def test_1darray(self, xp, dtype):
...
+class TestDiagIndices(unittest.TestCase):
+
+    @testing.numpy_cupy_array_equal()
+    def test_diag_indices(self, xp):
+        return xp.diag_indices(self.n, self.ndim)
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 3), (2, 2, 2)],
+}))
```
Comment (emcastillo) [inline tests/cupy_tests/indexing_tests/test_insert.py]: Test non-squared matrices and invalid values

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 3 above)
Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: I think there are no invalid values for this test, any pair of numbers would work fine.

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 4 above)
Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: The last two tests in this class, `test_non_equal_dims` and `test_1darray` are to check for the invalid values.

File: cupy/indexing/insert.py (hunk 2 above)
Comment (rushabh-v) [reply cupy/indexing/insert.py]: Done

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 4 above)
Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Or do you mean something else by `invalid values`?

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 3 above)
Comment (emcastillo) [reply tests/cupy_tests/indexing_tests/test_insert.py]: negative values for `n` and `ndim`, unidimensional cases, etc, ...
values that should make the routine fail or give an error.
Also corner cases such as 0, dim matrix etc

Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Those cases also don't fail or give any error on NumPy (They return empty tuples and so on.) But yes, we should definitely add those cases in the tests :+1:

Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: added those cases!

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 4 above)
Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Are those tests okay?

**Comment by rushabh-v** [comment]
I think we do not need to do benchmark here, right?

**Comment by rushabh-v** [comment]
Have rewritten the tests to not use of `testing.numpy_cupy_raises()` as @asi1024 is departing it with #3098 .

File: tests/cupy_tests/indexing_tests/test_insert.py [hunk 5]
```diff
@@ -155,3 +155,39 @@ def test_1darray(self, dtype):
...
+        for xp in (numpy, cupy):
+            arr = testing.shaped_arange((3, 5), xp)
+            with pytest.raises(ValueError):
+                xp.diag_indices_from(arr)
+
+    def test_1darray(self):
+        for xp in (numpy, cupy):
+            arr = testing.shaped_arange((5,), xp)
+            with pytest.raises(ValueError):
+                xp.diag_indices_from(arr)
```
Comment (rushabh-v) [inline tests/cupy_tests/indexing_tests/test_insert.py]: Should we write these tests in a separate class to remove unnecessary multiple runs of these tests while testing? I mean they don't use `self.shape` but still will run twice in this particular case?

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 3 above)
Comment (emcastillo) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Please make a specific class for invalid tests

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 4 above)
Comment (emcastillo) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Shapes like (0,0), (3,4), (5), (-1) and so

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 5 above)
Comment (emcastillo) [reply tests/cupy_tests/indexing_tests/test_insert.py]: Yes

**Comment by rushabh-v** [comment]
I have added new test cases.

File: cupy/indexing/insert.py [hunk 6]
```diff
@@ -118,3 +118,75 @@ def fill_diagonal(a, val, wrap=False):
...
+    Return the indices to access the main diagonal of an n-dimensional array.
+    See `diag_indices` for full details.
+
+    Args:
+        arr (cupy.ndarray): At least 2-D.
+
+     .. seealso:: :func:`numpy.diag_indices_from`
+
+    """
+    if not arr.ndim >= 2:
```
Comment (emcastillo) [inline cupy/indexing/insert.py:185]: Check that arr is a cupy array.

Something like
```python
if not isinstance(arr, cupy.ndarray):
    raise TypeError('Argument must be cupy.ndarray')
```

Comment (rushabh-v) [reply cupy/indexing/insert.py:178]: Done.

Comment (rushabh-v) [reply cupy/indexing/insert.py:178]: And the merge conflicts are also fixed.

File: tests/cupy_tests/indexing_tests/test_insert.py (hunk 3 above)
Comment (emcastillo) [reply tests/cupy_tests/indexing_tests/test_insert.py:163]: Please make a specific class for invalid tests

Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py:163]: Oh, I am sorry, I didn't notice your earlier comment until now. I have separated those cases.

Comment (rushabh-v) [reply tests/cupy_tests/indexing_tests/test_insert.py:163]: can you take a look, please? @emcastillo

**Comment by emcastillo** [comment]
Jenkins, test this please
